### PR TITLE
Add a context menu to rename or delete a heap snapshot from the list

### PR DESCRIPTION
**Pull request by kenzieschmoll**
Fixes https://github.com/flutter/devtools/issues/5954. 

rename-or-delete

### CODE REVIEW

**Comment by polina-c** [comment]
Thanks for putting it together. This way the feature does not look to be discoverable. Why do not we add icon with pen to the selected snapshot? Or, we can add context menu (vertical `...`) with two items: Delete, Rename.

**Comment by kenzieschmoll** [comment]
A user may want to edit a snapshot name even if it is not selected, so for this reason, having an icon that only appears for the selected snapshot may also have usability issues. Adding an overflow menu (vertical `...`) for each snapshot item could work, though the UI may look busy with that icon on each row.

I would argue that right click is discoverable - double click was admittedly a bonus item that users may or may not find.
CC @InMatrix

**Comment by polina-c** [comment]
Renaming is not needed very often. So, I think it is ok to show the icon just for selected row, the same way as we do it for table:

Screenshot 2023-07-06 at 2 54 18 PM

Right click is discoverable in general, but not in DevTools, where we do not utilize it a lot. Even if user discovers it, they will not expect second row to become editable after selecting 'Rename' for first row.

Also, implementation with  vertical `...` will be conciser in code, and thus cheaper to support.

**Comment by kenzieschmoll** [comment]
> they will not expect second row to become editable after selecting 'Rename' for first row.

This is not the case. Only the row that was right clicked will become editable. 

I will look into adding the overflow menu icon for the selected row, but I don't see an issue with also leaving the right click functionality. We haven't used right click much in the past because flutter's context menu support for web wasn't as good as it is now.

**Comment by polina-c** [comment]
> I will look into adding the overflow menu icon for the selected row, but I don't see an issue with also leaving the right click functionality. We haven't used right click much in the past because flutter's context menu support for web wasn't as good as it is now. 

Would it be possible to incapsulate right click handling into mixin somehow? So that it does not crowd the code of the widget, where it is used? 

BTW, in modern cross-plat UI, context actions are more and more behind vertical ellipsis.
So, it is becoming user habit to assume there is no actions, if there is no ellipsis:

Screenshot 2023-07-07 at 8 31 44 AM

**Comment by polina-c** [comment]
> This is not the case. Only the row that was right clicked will become editable.

On the attached video I see click on 'Rename' just one time, and then I see two snapshots are renamed. How it is?

**Comment by kenzieschmoll** [comment]
> On the attached video I see click on 'Rename' just one time, and then I see two snapshots are renamed. How it is?

This was from double clicking in the second row, which I've since removed.

**Comment by kenzieschmoll** [comment]
Okay this is ready for another review. I removed the right click and double click functionality in favor of the context menu. The benefit of maintaining both affordances is not worth the cost. Updated the PR description.

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_list.dart [hunk 1]
```diff
@@ -106,52 +107,195 @@ class _ListControlPane extends StatelessWidget {
...
+    required this.onDelete,
+  }) : super(key: key);
 
   final SnapshotItem item;
 
+  final int index;
+
+  final bool selected;
+
+  final ValueListenable<int?> editIndex;
```
Comment (polina-c) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_list.dart:128]: Can you add doc comment for the new members. Some are not obvious (editIndex, onEdit vs onEditComplete)

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_list.dart:129]: done

**Review by polina-c — APPROVED** [review]
LGTM after comments

**Pull request merged**